### PR TITLE
FLPATH-313: get workflow by projectId

### PR DESCRIPTION
**Pull request by RichardW98**
[FLPATH-313](https://issues.redhat.com//browse/FLPATH-313): get workflow by projectId

**Change type**
- [ ] New feature
- [ ] Bug fix
- [ ] Unit tests
- [ ] Integration tests
- [ ] CI
- [ ] Documentation
- [ ] Auto generated SDK code

**Impacted services**
- [ ] Workflow Service
- [ ] Notifivcation Service

**Checklist**
- [ ] Subject and description added to both, commit and PR.
- [ ] Relevant issues have been referenced.

### CODE REVIEW

File: workflow-service/src/test/java/com/redhat/parodos/workflow/execution/controller/WorkFlowControllerTest.java [hunk 1]
```diff
@@ -231,6 +234,22 @@ public void TestGetStatusWithValidData() throws Exception {
 				.andExpect(MockMvcResultMatchers.status().isOk());
 	}
 
+	@Test
+	void TestGetStatusByProjectIdWithValidData() throws Exception {
```
Comment (eloycoto) [inline workflow-service/src/test/java/com/redhat/parodos/workflow/execution/controller/WorkFlowControllerTest.java:238]: I miss the following test:


- where checks that the project is nill? Does it return all projects?
- What happens is getWorkFlowStatusByProjectId return an empty list?

File: workflow-service/src/test/java/com/redhat/parodos/workflow/execution/service/WorkFlowServiceImplTest.java [hunk 2]
```diff
@@ -907,6 +909,27 @@ public void testGetWorkflowParametersWithoutWorkflowOptions() {
 		assertNull(workflowParameters.getWorkFlowOptions().getNewOptions());
 	}
 
+	@Test
+	void getWorkFlowStatusByProjectId_when_projectIsFound_then_returnWorkFlowStatus() {
```
Comment (eloycoto) [inline workflow-service/src/test/java/com/redhat/parodos/workflow/execution/service/WorkFlowServiceImplTest.java:913]: We need to increase testing when no workflows with valid projects and what happens when JPA repo returns nil

File: workflow-service/src/main/java/com/redhat/parodos/workflow/execution/service/WorkFlowServiceImpl.java [hunk 3]
```diff
@@ -173,6 +173,12 @@ public synchronized WorkFlowExecution updateWorkFlow(WorkFlowExecution workFlowE
 		return workFlowRepository.save(workFlowExecution);
 	}
 
+	@Override
+	public List<WorkFlowStatusResponseDTO> getWorkFlowStatusByProjectId(UUID projectId) {
+		return workFlowRepository.findAllByProjectId(projectId).stream()
+				.map(workFlowExecution -> getWorkFlowStatus(workFlowExecution.getId())).toList();
```
Comment (eloycoto) [inline workflow-service/src/main/java/com/redhat/parodos/workflow/execution/service/WorkFlowServiceImpl.java:179]: Why is the reason for this map?
Could you add a test for this?

**Review by eloycoto — CHANGES_REQUESTED** [review]

File: workflow-service/src/test/java/com/redhat/parodos/workflow/execution/controller/WorkFlowControllerTest.java [hunk 4]
```diff
@@ -231,6 +234,22 @@ public void TestGetStatusWithValidData() throws Exception {
...
+		Mockito.when(workFlowService.getWorkFlowStatusByProjectId(projectId))
+				.thenReturn(List.of(WorkFlowStatusResponseDTO.builder().workFlowName("test-workflow")
+						.status(ParodosWorkStatus.COMPLETED.name()).build()));
+		// when
+		this.mockMvc
+				.perform(this.getRequestWithValidCredentials("/api/v1/workflows/status").param("projectId",
+						projectId.toString()))
+				.andExpect(MockMvcResultMatchers.status().isOk())
+				.andExpect(MockMvcResultMatchers.content().contentType(MediaType.APPLICATION_JSON))
+				.andExpect(MockMvcResultMatchers.jsonPath("$", hasSize(1))).andExpect(MockMvcResultMatchers
```
Comment (gciavarrini) [inline workflow-service/src/test/java/com/redhat/parodos/workflow/execution/controller/WorkFlowControllerTest.java:249]: not sure if a formatter issue: can you move the last `andExpect` to a new line?

File: workflow-service/src/main/java/com/redhat/parodos/workflow/execution/service/WorkFlowServiceImpl.java (hunk 3 above)
Comment (gciavarrini) [reply workflow-service/src/main/java/com/redhat/parodos/workflow/execution/service/WorkFlowServiceImpl.java:179]: @eloycoto `stream().map()` allows applying a function to each stream element. So for me, it makes sense to iterate over the element and call `getWorkflowStatus` and then collect all results into a List. 
However, I agree that a unit test for this method is valuable.

File: workflow-service/src/main/java/com/redhat/parodos/workflow/execution/controller/WorkFlowController.java [hunk 5]
```diff
@@ -124,6 +124,20 @@ public ResponseEntity<WorkFlowStatusResponseDTO> getStatus(@PathVariable String
...
+	@ApiResponses(
+			value = {
+					@ApiResponse(responseCode = "200", description = "Succeeded",
+							content = { @Content(mediaType = "application/json",
+									schema = @Schema(implementation = List.class)) }),
+					@ApiResponse(responseCode = "401", description = "Unauthorized", content = @Content),
+					@ApiResponse(responseCode = "403", description = "Forbidden", content = @Content) })
+	@GetMapping("/status")
+	public ResponseEntity<List<WorkFlowStatusResponseDTO>> getStatusByProjectId(
+			@RequestParam("projectId") String projectId) {
```
Comment (masayag) [inline workflow-service/src/main/java/com/redhat/parodos/workflow/execution/controller/WorkFlowController.java:137]: projectId should be marked as @NotNull

File: workflow-service/src/main/java/com/redhat/parodos/workflow/execution/controller/WorkFlowController.java [hunk 6]
```diff
@@ -124,6 +124,20 @@ public ResponseEntity<WorkFlowStatusResponseDTO> getStatus(@PathVariable String
...
+	@Operation(summary = "Returns workflows' status by project id")
+	@ApiResponses(
+			value = {
+					@ApiResponse(responseCode = "200", description = "Succeeded",
+							content = { @Content(mediaType = "application/json",
+									schema = @Schema(implementation = List.class)) }),
+					@ApiResponse(responseCode = "401", description = "Unauthorized", content = @Content),
+					@ApiResponse(responseCode = "403", description = "Forbidden", content = @Content) })
+	@GetMapping("/status")
+	public ResponseEntity<List<WorkFlowStatusResponseDTO>> getStatusByProjectId(
```
Comment (masayag) [inline workflow-service/src/main/java/com/redhat/parodos/workflow/execution/controller/WorkFlowController.java:136]: not sure what is the reason for it, but the generated openapi.json returns a string as a result while I'd expect the type of the return value to be a list.

however, if the project is mandatory in this context, I wonder why not expose the project-status endpoint instead.

if the purpose is to evaluate the project status, then it seems more natural to query for the project resource instead of the workflow, as project status might contain other information relevant to the project's overall status.

What will be the use of the new endpoint?

Comment (gciavarrini) [reply workflow-service/src/main/java/com/redhat/parodos/workflow/execution/controller/WorkFlowController.java:136]: Since this method return the list of the statuses of all the workflows available in a project (if I understood correctly), maybe renaming the method into `getStatusesByProjectId` can ease the understanding.

File: workflow-service/src/test/java/com/redhat/parodos/workflow/execution/controller/WorkFlowControllerTest.java (hunk 1 above)
Comment (RichardW98) [reply workflow-service/src/test/java/com/redhat/parodos/workflow/execution/controller/WorkFlowControllerTest.java:238]: 1. yes  it should return all projects which the user is allowed to check
2. if it returns empty list then UI won't show any executions in the workflow table

Comment (RichardW98) [reply workflow-service/src/test/java/com/redhat/parodos/workflow/execution/controller/WorkFlowControllerTest.java:238]: I will add tests for them

File: workflow-service/src/main/java/com/redhat/parodos/workflow/execution/service/WorkFlowServiceImpl.java (hunk 3 above)
Comment (RichardW98) [reply workflow-service/src/main/java/com/redhat/parodos/workflow/execution/service/WorkFlowServiceImpl.java:179]: we have unit tests for `getWorkFlowStatus()`: https://github.com/RichardW98/parodos/blob/filter-workflow-by-projectid/workflow-service/src/test/java/com/redhat/parodos/workflow/execution/service/WorkFlowServiceImplTest.java#L444

File: workflow-service/src/main/java/com/redhat/parodos/workflow/execution/controller/WorkFlowController.java (hunk 6 above)
Comment (RichardW98) [reply workflow-service/src/main/java/com/redhat/parodos/workflow/execution/controller/WorkFlowController.java:136]: @gciavarrini @masayag new UI design has replaced the project Overview with workflows Overview.  Therefore project-status will not be needed any more, and instead we will need to show all workflow executions' info by project. I will update this pr to include info other than status in the return (e.g. startTeim,  endTime ...)

File: workflow-service/src/test/java/com/redhat/parodos/workflow/execution/controller/WorkFlowControllerTest.java (hunk 4 above)
Comment (RichardW98) [reply workflow-service/src/test/java/com/redhat/parodos/workflow/execution/controller/WorkFlowControllerTest.java:249]: yea. it's formatter 😅

File: workflow-service/src/main/java/com/redhat/parodos/project/repository/ProjectRepository.java [hunk 7]
```diff
@@ -26,4 +29,8 @@
  */
 public interface ProjectRepository extends JpaRepository<Project, UUID> {
 
+	List<Project> findAllByUserUsername(String username);
```
Comment (masayag) [inline workflow-service/src/main/java/com/redhat/parodos/project/repository/ProjectRepository.java:32]: pls add tests.
should be fairly simple with https://github.com/parodos-dev/parodos/pull/295/files#diff-43f3718db35688f81c93147df5fad5e6caf4bbdc988a24d007edd32db7a89372 (once it is being merged)

Comment (RichardW98) [reply workflow-service/src/main/java/com/redhat/parodos/project/repository/ProjectRepository.java:32]: yes Thank you for the example ~

**Comment by RichardW98** [comment]
I renamed 'user' table to 'users' temporarily, will fix it in another pr to standandize table names

File: workflow-service/generated/openapi/openapi.json [hunk 8]
```diff
@@ -259,6 +259,40 @@
       }
     },
     "/api/v1/workflows" : {
+      "get" : {
+        "operationId" : "getStatusByProjectId",
+        "parameters" : [ {
+          "in" : "query",
+          "name" : "projectId",
+          "required" : false,
```
Comment (masayag) [inline workflow-service/generated/openapi/openapi.json:267]: that should have rendered as `required`

File: workflow-service/src/main/java/com/redhat/parodos/workflow/execution/service/WorkFlowServiceImpl.java [hunk 9]
```diff
@@ -172,6 +185,26 @@ public synchronized WorkFlowExecution updateWorkFlow(WorkFlowExecution workFlowE
 		return workFlowRepository.save(workFlowExecution);
 	}
 
+	@Override
+	public List<WorkFlowResponseDTO> getWorkFlowsByProjectId(UUID projectId) {
+		List<ProjectResponseDTO> projects = projectId == null
+				? projectService.findProjectsByUserName(SecurityUtils.getUsername())
+				: List.of(projectService.getProjectByIdAndUserName(projectId, SecurityUtils.getUsername()));
+		return projects.stream().flatMap(project -> workFlowRepository.findAllByProjectId(project.getId()).stream()
+				.filter(workFlowExecution -> workFlowExecution.getMainWorkFlowExecution() == null)
```
Comment (masayag) [inline workflow-service/src/main/java/com/redhat/parodos/workflow/execution/service/WorkFlowServiceImpl.java:194]: a lot is going on in this method, but in terms of readability, I find its readability being somewhat complex.
do you think it can be refactored a bit?

File: workflow-service/generated/openapi/openapi.json (hunk 8 above)
Comment (RichardW98) [reply workflow-service/generated/openapi/openapi.json:267]: I changed it to false in the controller, so that we can return all executed workflows for this user

File: workflow-service/src/main/java/com/redhat/parodos/workflow/execution/service/WorkFlowServiceImpl.java [hunk 10]
```diff
@@ -172,6 +185,26 @@ public synchronized WorkFlowExecution updateWorkFlow(WorkFlowExecution workFlowE
 		return workFlowRepository.save(workFlowExecution);
 	}
 
+	@Override
+	public List<WorkFlowResponseDTO> getWorkFlowsByProjectId(UUID projectId) {
+		List<ProjectResponseDTO> projects = projectId == null
```
[the remark this answers is not in the document]
Comment (RichardW98) [reply workflow-service/src/main/java/com/redhat/parodos/workflow/execution/service/WorkFlowServiceImpl.java:190]: yes then this controller will return all the workflows for this user, not limited to certain projects

File: workflow-service/src/main/java/com/redhat/parodos/workflow/execution/service/WorkFlowServiceImpl.java (hunk 9 above)
Comment (RichardW98) [reply workflow-service/src/main/java/com/redhat/parodos/workflow/execution/service/WorkFlowServiceImpl.java:194]: ok, I will move out the WorkFlowResponseDTO builder

File: workflow-service/src/main/java/com/redhat/parodos/security/SecurityUtils.java [hunk 11]
```diff
@@ -40,13 +41,15 @@ private SecurityUtils() {
...
-		Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
-		if (authentication instanceof JwtAuthenticationToken) {
-			return ((JwtAuthenticationToken) authentication).getToken().getClaim("preferred_username");
+	public static String getUsername() {
+		LdapUserDetailsImpl ldapDetails = null;
+		try {
+			ldapDetails = (LdapUserDetailsImpl) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
+			return ldapDetails.getUsername();
+		}
+		catch (NullPointerException e) {
```
[the remark this answers is not in the document]
Comment (RichardW98) [reply workflow-service/src/main/java/com/redhat/parodos/security/SecurityUtils.java:49]: SecurityContextHolder.getContext().getAuthentication() might be null. I will change it to if/else

File: workflow-service/src/main/java/com/redhat/parodos/workflow/execution/service/WorkFlowServiceImpl.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (RichardW98) [reply workflow-service/src/main/java/com/redhat/parodos/workflow/execution/service/WorkFlowServiceImpl.java:190]: ok I will separate them

File: workflow-service/src/main/java/com/redhat/parodos/workflow/execution/dto/WorkFlowResponseDTO.java [hunk 12]
```diff
@@ -36,14 +36,23 @@
 @Data
 @AllArgsConstructor
 @NoArgsConstructor
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
 public class WorkFlowResponseDTO {
 
 	private UUID workFlowExecutionId;
 
-	@JsonInclude(JsonInclude.Include.NON_EMPTY)
+	private String projectId;
```
Comment (masayag) [inline workflow-service/src/main/java/com/redhat/parodos/workflow/execution/dto/WorkFlowResponseDTO.java:44]: pls change to UUID.

File: workflow-service/src/test/java/com/redhat/parodos/project/service/ProjectServiceImplTest.java [hunk 13]
```diff
@@ -65,18 +76,30 @@ public void testFindProjectByIdWithValidData() {
...
 		// when
-		Exception exception = assertThrows(RuntimeException.class, () -> {
-			this.projectService.getProjectById(project.getId());
-		});
+		ProjectResponseDTO res = this.projectService.getProjectByIdAndUsername(project.getId(), username);
 
 		// then
-		assertEquals(exception.getMessage(), String.format("Project with id: %s not found", project.getId()));
+		assertNotNull(res);
+		assertEquals(res.getId().toString(), project.getId().toString());
```
Comment (masayag) [inline workflow-service/src/test/java/com/redhat/parodos/project/service/ProjectServiceImplTest.java:91]: toString() can be removed from both sides

**Review by masayag — APPROVED** [review]
